{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook for creating TableS12, TableS18 and making a preparation for feature matrix (Table 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Configutation part #\n",
    "\n",
    "# Directory with matrix after deduplication\n",
    "data_dir = \"../db/cv_bess_files\"\n",
    "# Directory with matrix before deduplication\n",
    "data_dir_pre = \"../db/cv_bess_files.2\"\n",
    "# Directory with output of hhs_lr.py\n",
    "output_hhs_lr = './output_hhs_lr'\n",
    "# Output directory\n",
    "final_output_hhs = './final_output_hhs'\n",
    "# Annotation file\n",
    "annotation_file = '../db/AL123456.gff'\n",
    "# Resistance genes file\n",
    "resistance_gene_file = '../db/resistance_genes.csv'\n",
    "\n",
    "######################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "from copy import deepcopy\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from tqdm import tqdm\n",
    "\n",
    "from scipy.stats import fisher_exact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(final_output_hhs):\n",
    "    os.makedirs(final_output_hhs)\n",
    "    \n",
    "thr = '3'\n",
    "\n",
    "drugs = ['Kanamycin', 'Amikacin', 'Streptomycin', 'Ofloxacin',\n",
    "         'Moxifloxacin', 'Isoniazid', 'Rifampicin', 'Ethambutol',\n",
    "         'Pyrazinamide', 'Capreomycin', 'Ethionamide', 'Prothionamide',\n",
    "         'Ciprofloxacin']\n",
    "\n",
    "annotation = pd.read_csv(annotation_file, sep='\\t', index_col=None, header=None)\n",
    "annotation = annotation[annotation[2] == \"CDS\"]\n",
    "annotation['gene'] = [x.split(';')[2].split()[-1] for x in annotation[8]]\n",
    "annotation['seq'] = [x.split(';')[-1].split('translation')[-1].replace(' ','').replace('\"', '') for x in annotation[8]]\n",
    "annotation = annotation[['gene', 'seq']]\n",
    "\n",
    "res_data = pd.read_csv(resistance_gene_file, sep='\\t', \n",
    "                               index_col=None, header=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Support functions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def represents_int(s):\n",
    "    try: \n",
    "        int(s)\n",
    "        return True\n",
    "    except ValueError:\n",
    "        return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def conj_matrix(X):\n",
    "    value00 = len(X[(X.iloc[:,0] == 0) & (X.iloc[:,1] == 0)])\n",
    "    value01 = len(X[(X.iloc[:,0] == 0) & (X.iloc[:,1] == 1)])\n",
    "    value10 = len(X[(X.iloc[:,0] == 1) & (X.iloc[:,1] == 0)])\n",
    "    value11 = len(X[(X.iloc[:,0] == 1) & (X.iloc[:,1] == 1)])\n",
    "    return [value00, value01, value10, value11]\n",
    "\n",
    "def fisher(X):\n",
    "    n = len(X)\n",
    "    [value00, value01, value10, value11] = conj_matrix(X)\n",
    "    return fisher_exact([[value00, value01], [value10, value11]])[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ind_by_name(pos_info, name, drug):\n",
    "    if name.split('_')[-1] == 'domain':\n",
    "        gene = \"_\".join(name.split('_')[:-1])\n",
    "        temp = pos_info[(pos_info['gene'] == gene) & (pos_info['pos'] == 'changed')]\n",
    "    elif name.split('_')[-1] == 'broken':\n",
    "        gene = \"_\".join(name.split('_')[:-1])\n",
    "        temp = pos_info[(pos_info['gene'] == gene) & (pos_info['pos'] == 'broken')]\n",
    "    elif name.split('_')[-1] == 'snp':\n",
    "        gene = \"_\".join(name.split('_')[:-3])\n",
    "        pos = name.split('_')[-3]\n",
    "        ind = name.split('_')[-2]\n",
    "        temp = pos_info[(pos_info['gene'] == gene) & (pos_info['pos'] == pos) & (pos_info['ind2'] == ind) & (pos_info['act'] == 'snp')]\n",
    "    elif name.split('_')[-1] == 'del':\n",
    "        gene = \"_\".join(name.split('_')[:-3])\n",
    "        pos = name.split('_')[-3]\n",
    "        ind = name.split('_')[-2]\n",
    "        if represents_int(ind):\n",
    "            temp = pos_info[(pos_info['gene'] == gene) & (pos_info['pos'] == pos) & (pos_info['ind1'] == ind)  & (pos_info['ind2'] == 'del')]\n",
    "        else:\n",
    "            temp = pos_info[(pos_info['gene'] == gene) & (pos_info['pos'] == pos) & (pos_info['ind2'] == ind)  & (pos_info['act'] == 'del')]\n",
    "    elif name.split('_')[-1] == 'ins':\n",
    "        gene = \"_\".join(name.split('_')[:-3])\n",
    "        pos = name.split('_')[-3]\n",
    "        ind = name.split('_')[-2]\n",
    "        temp = pos_info[(pos_info['gene'] == gene) & (pos_info['pos'] == pos) & (pos_info['ind2'] == ind)  & (pos_info['act'] == 'ins')]\n",
    "\n",
    "    if len(temp) == 0: \n",
    "        return -1\n",
    "    else: \n",
    "        return temp.index[0]\n",
    "    \n",
    "def get_name_by_ind(pos_info, j, drug):\n",
    "    temp = pos_info.iloc[j]\n",
    "    gene, pos, ind1, ind2, act = temp['gene'], temp['pos'], temp['ind1'], temp['ind2'], temp['act']\n",
    "\n",
    "    if pos == 'changed':\n",
    "        return f\"{gene}_domain\"\n",
    "    elif pos == 'broken':\n",
    "        return f\"{gene}_broken\"\n",
    "    elif act == 'snp':\n",
    "        return f\"{gene}_{pos}_{ind2}_snp\"\n",
    "    elif ind2 == 'del':\n",
    "        return f\"{gene}_{pos}_{ind1}_{ind2}\"\n",
    "    elif act == 'del':\n",
    "        return f\"{gene}_{pos}_{ind2}_{act}\"\n",
    "    elif act == 'ins':\n",
    "        return f\"{gene}_{pos}_{ind2}_{act}\"\n",
    "    \n",
    "def get_description(pos_info, ind):\n",
    "    gene, pos, fr, to, act = pos_info.loc[ind].values\n",
    "    if act == 'snp':\n",
    "        return [gene, pos, 'SNV', fr, to]\n",
    "    elif act == 'ins':\n",
    "        return [gene, pos, 'Insertion', fr, to]\n",
    "    elif act == 'del':\n",
    "        return [gene, pos, 'Deletion', fr, to]\n",
    "    elif to == 'del':\n",
    "        seq = annotation[annotation['gene'] == gene]['seq'].to_numpy()[0]\n",
    "        fr = seq[int(pos)-1: int(pos)+int(fr)-1]\n",
    "        return [gene, pos, 'Deletion', fr, \"-\"]\n",
    "    elif pos == 'changed':\n",
    "        temp = [i for i in gene.split('_') if not represents_int(i)]\n",
    "        new_gene = \"_\".join(temp[:-1])\n",
    "        domain = \"_\".join(gene.split('_')[len(new_gene.split('_')):])\n",
    "        return [new_gene, '-', f'{domain} is changed', '-', '-']\n",
    "        return f\"In {new_gene} PFAM {domain} is changed\"\n",
    "    elif pos == 'broken':\n",
    "        return [gene, '-', 'Broken', '-', '-']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HHS Metrics (TableS18)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "hhs_auc_1 = pd.read_csv(f\"{output_hhs_lr}/1/auc.csv\", \n",
    "                        sep='\\t', index_col=0, header=0)\n",
    "hhs_auc_1['filter'] = 1\n",
    "hhs_auc_3 = pd.read_csv(f\"{output_hhs_lr}/3/auc.csv\", \n",
    "                        sep='\\t', index_col=0, header=0)\n",
    "hhs_auc_3['filter'] = 3\n",
    "hhs_auc_5 = pd.read_csv(f\"{output_hhs_lr}/5/auc.csv\", \n",
    "                        sep='\\t', index_col=0, header=0)\n",
    "hhs_auc_5['filter'] = 5\n",
    "\n",
    "hhs_auc = pd.concat([hhs_auc_1, hhs_auc_3, hhs_auc_5], axis=0)\n",
    "\n",
    "hhs_auc = hhs_auc[hhs_auc.groupby(level=0)['auc'].transform(max) == hhs_auc['auc']]\n",
    "\n",
    "hhs_auc.to_csv(f\"./{final_output_hhs}/TableS18.csv\", sep='\\t', index=True, header=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add statical info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 13/13 [10:47<00:00, 49.79s/it]\n"
     ]
    }
   ],
   "source": [
    "hhs_auc = pd.read_csv(f\"./{final_output_hhs}/TableS18.csv\", \n",
    "                      sep='\\t', index_col=0, header=0)\n",
    "\n",
    "for drug in tqdm(drugs):\n",
    "    f = hhs_auc.loc[drug, 'filter']\n",
    "    \n",
    "    pos_info = pd.read_csv(f\"{data_dir_pre}/{drug}.gt.pos.domains.3.0\", sep='\\t', \n",
    "                              index_col=None, header=None, usecols=[0, 1, 2, 3, 4])\n",
    "    pos_info.columns = ['gene', 'pos', 'ind1', 'ind2', 'act']\n",
    "    \n",
    "    for i in range(5):\n",
    "        pos_info_temp = pd.read_csv(f\"{data_dir_pre}/{drug}.gt.pos.domains.3.{i}\", sep='\\t', \n",
    "                              index_col=None, header=None, usecols=[0, 1, 2, 3, 4])\n",
    "        pos_info_temp.columns = ['gene', 'pos', 'ind1', 'ind2', 'act']\n",
    "        \n",
    "        temp = pd.read_csv(f\"{output_hhs_lr}/{f}/{drug}.gt.domains.3.final.{i}\", \n",
    "                           sep='\\t', index_col=0, header=0)\n",
    "        indeces = [int(str(x).split(', ')[0]) for x in temp.index]\n",
    "        indeces = [get_ind_by_name(pos_info, get_name_by_ind(pos_info_temp, i, drug), drug) for i in indeces]\n",
    "        \n",
    "        sel_indeces = deepcopy(indeces)\n",
    "        while -1 in sel_indeces: sel_indeces.remove(-1)\n",
    "        \n",
    "        X_test = pd.read_csv(f\"{data_dir_pre}/{drug}.gt.domains.3.test.0\", sep=' ', \n",
    "                             index_col=None, header=None, usecols=sel_indeces)\n",
    "        X_train = pd.read_csv(f\"{data_dir_pre}/{drug}.gt.domains.3.train.0\", sep=' ', \n",
    "                             index_col=None, header=None, usecols=sel_indeces)\n",
    "        X = pd.concat([X_train, X_test])\n",
    "        \n",
    "        y_train= pd.read_csv(f\"{data_dir_pre}/{drug}.phen.domains.3.train.0\", sep=' ', \n",
    "                             index_col=None, header=None)\n",
    "        y_test = pd.read_csv(f\"{data_dir_pre}/{drug}.phen.domains.3.test.0\", sep=' ', \n",
    "                             index_col=None, header=None)\n",
    "        y = pd.concat([y_train, y_test])\n",
    "        X['y'] = y.to_numpy().ravel()\n",
    "\n",
    "        pvalues = []\n",
    "        values00 = []\n",
    "        values01 = []\n",
    "        values10 = []\n",
    "        values11 = []\n",
    "\n",
    "        for j in indeces:\n",
    "            if j == -1: \n",
    "                pvalues.append('-')\n",
    "                [value00, value01, value10, value11] = ['-', '-', '-', '-']\n",
    "            else: \n",
    "                pvalues.append(fisher(X[[j, 'y']]))\n",
    "                [value00, value01, value10, value11] = conj_matrix(X[[j, 'y']])\n",
    "            values00.append(value00)\n",
    "            values01.append(value01)\n",
    "            values10.append(value10)\n",
    "            values11.append(value11)\n",
    "        temp['0_0'] = values00\n",
    "        temp['0_1'] = values01\n",
    "        temp['1_0'] = values10\n",
    "        temp['1_1'] = values11\n",
    "        temp['pvalue'] = pvalues\n",
    "        temp.to_csv(f\"{output_hhs_lr}/{f}/{drug}.gt.domains.3.final.{i}\", \n",
    "                    sep='\\t', header=True, index=True)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add aggregated info file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "for drug in drugs: \n",
    "    f = hhs_auc.loc[drug, 'filter'] \n",
    "    \n",
    "    input_dir = f\"./{output_hhs_lr}/{f}\"\n",
    "    output_dir = f\"./{output_hhs_lr}/{f}\"\n",
    "    \n",
    "    all_data = pd.DataFrame(columns=['ind', 'name', 'coef'])\n",
    "    count_features = pd.DataFrame()\n",
    "    for i in range(5):\n",
    "        pos_info = pd.read_csv(f\"{data_dir_pre}/{drug}.gt.pos.domains.3.{i}\", sep='\\t', \n",
    "                                  index_col=None, header=None, usecols=[0, 1, 2, 3, 4])\n",
    "        pos_info.columns = ['gene', 'pos', 'ind1', 'ind2', 'act']\n",
    "        pos_info.fillna('Nope', inplace=True)\n",
    "        \n",
    "        \n",
    "        temp = pd.read_csv(f\"{input_dir}/{drug}.gt.domains.{thr}.final.{i}\", \n",
    "                           sep='\\t', index_col=0, header=0)\n",
    "        \n",
    "        description = [\", \".join([get_name_by_ind(pos_info, int(i), drug) \n",
    "                                 for i in str(x).split(\", \")]) for x in temp.index]\n",
    "        \n",
    "        temp['Event description'] = description\n",
    "        temp.drop('name', axis=1)\n",
    "        \n",
    "        count_features = pd.concat([count_features, temp], axis=0)\n",
    "    count_features['count'] = 1  \n",
    "\n",
    "    count_features = count_features.groupby(by=['Event description']).agg({'coef': np.mean, \n",
    "                                                            'count': np.sum, \n",
    "                                                            'pvalue': lambda x: x.iloc[0]})\n",
    "\n",
    "    count_features.sort_values(by='count', inplace=True, ascending=False)\n",
    "\n",
    "    count_features.to_csv(f\"{output_dir}/count_{drug}_{thr}.csv\", sep='\\t', index=True, header=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Feature table HHS (TableS12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "hhs_auc = pd.read_csv(f\"./{final_output_hhs}/TableS18.csv\", \n",
    "                      sep='\\t', index_col=0, header=0)\n",
    "\n",
    "for drug in drugs:\n",
    "    filt = hhs_auc.loc[drug, 'filter']\n",
    "\n",
    "    count_features = pd.DataFrame()\n",
    "    result = pd.DataFrame(columns=['gene', 'pos', 'act', 'from', 'to', 'num', \n",
    "                                   'name', 'coef', 'pvalue', '0_0', \n",
    "                                       '0_1', '1_0', '1_1', 'count'])\n",
    "    for i in range(5):\n",
    "        temp = pd.read_csv(f\"{output_hhs_lr}/{filt}/{drug}.gt.domains.3.final.{i}\", \n",
    "                           sep='\\t', index_col=0, header=0)\n",
    "\n",
    "        count_features = pd.concat([count_features, temp], axis=0)\n",
    "    count_features['count'] = 1    \n",
    "\n",
    "    count_features = count_features.groupby(['name']).agg({'coef': np.mean, \n",
    "                                                                  'count': np.sum,\n",
    "                                                                  'pvalue': lambda x: x.iloc[0],\n",
    "                                                                  '0_0': lambda x: x.iloc[0],\n",
    "                                                                  '0_1': lambda x: x.iloc[0],\n",
    "                                                                  '1_0': lambda x: x.iloc[0],\n",
    "                                                                  '1_1': lambda x: x.iloc[0]})\n",
    "    \n",
    "    count_features.sort_values('count', ascending=False, inplace=True)\n",
    "\n",
    "    for j in range(len(count_features)):\n",
    "        names = count_features.index[j]\n",
    "        coef, count, pvalue, value00, value01, value10, value11 = count_features.to_numpy()[j]\n",
    "        for name in names.split(', '):\n",
    "            for i in range(5):\n",
    "                pos_info = pd.read_csv(f\"{data_dir_pre}/{drug}.gt.pos.domains.3.{i}\", sep='\\t', \n",
    "                                  index_col=None, header=None, usecols=[0, 1, 2, 3, 4])\n",
    "                pos_info.columns = ['gene', 'pos', 'ind1', 'ind2', 'act']\n",
    "                pos_info.fillna('Nope', inplace=True)\n",
    "                ind = get_ind_by_name(pos_info, name, drug)\n",
    "                if ind != -1:\n",
    "                    break\n",
    "                    \n",
    "            add = pd.DataFrame([[*get_description(pos_info, ind), j, name, coef, pvalue, value00,\n",
    "                                 value01, value10, value11, count]], \n",
    "                              columns=['gene', 'pos', 'act', 'from', 'to', \n",
    "                                       'num', 'name', 'coef', 'pvalue', '0_0', \n",
    "                                       '0_1', '1_0', '1_1', 'count'])\n",
    "            result = pd.concat([result, add])\n",
    "    result.drop(['name'], axis=1, inplace=True)\n",
    "    result['drug'] = drug\n",
    "    result.to_csv(f\"{final_output_hhs}/{drug}.csv\", sep='\\t', index=False, header=True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_data = pd.DataFrame(columns=['drug', 'gene', 'pos', 'act', 'from', 'to', \n",
    "                                       'num', 'coef', 'pvalue', '0_0', \n",
    "                                       '0_1', '1_0', '1_1', 'count'])\n",
    "for drug in drugs:\n",
    "    result = pd.read_csv(f\"{final_output_hhs}/{drug}.csv\", sep='\\t', index_col=None, header=0)\n",
    "    result['drug'] = drug\n",
    "    all_data = pd.concat([all_data, result], axis=0)\n",
    "all_data.to_csv(f\"{final_output_hhs}/TableS12.csv\", sep='\\t', header=None, index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HHS feature metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "hhs_result = pd.DataFrame(index=drugs, \n",
    "                            columns=['res_gen', 'jacard', \n",
    "                                     'sel_true_gene', 'sel_false_gene'])\n",
    "for drug in drugs:\n",
    "    f = hhs_auc.loc[drug, 'filter']\n",
    "    features = pd.read_csv(f\"{output_hhs_lr}/{f}/count_{drug}_3.csv\", \n",
    "                       sep='\\t', index_col=0, header=0)\n",
    "    features = features[features['count'] >= 3]\n",
    "\n",
    "    sel_genes = list(set([x.split('_')[0] for x in features.index]))\n",
    "    sel_genes = pd.DataFrame(sel_genes, columns=['gene'])\n",
    "    sel_genes['true'] = sel_genes['gene'].isin(res_data[res_data['drug'] == drug]['gene'])\n",
    "    sel_genes['temp'] = sel_genes['gene'].isin(res_data[res_data['drug'] != drug]['gene'])\n",
    "    sel_genes['false'] = ~sel_genes['true'] & sel_genes['temp']\n",
    "\n",
    "    res_gene = len(res_data[res_data['drug'] == drug])\n",
    "    sel_true_gene = np.sum(sel_genes['true'])\n",
    "    sel_false_gene = np.sum(sel_genes['false'])\n",
    "    jacard = sel_true_gene/(res_gene + len(features) - sel_true_gene)\n",
    "\n",
    "    hhs_result.loc[drug] = [res_gene, jacard, sel_true_gene, sel_false_gene]\n",
    "    \n",
    "hhs_result.to_csv(f\"{final_output_hhs}/hhs_feature_metrics.csv\", sep='\\t',\n",
    "                 index=True, header=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
