{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook to create quality prediction by direct association method using WHO dictonary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Configutation part #\n",
    "\n",
    "# Directory with nucleotide substitution files \n",
    "nucl_data_dir = '../db/nucl_data'\n",
    "# Directory with phenotypes\n",
    "pheno_dir = '../../pheno'\n",
    "# Directory with matrices without split to parts and without excluding rare mutations\n",
    "data_dir = ''\n",
    "# WHO catalogue\n",
    "who_catalogue_file = \"../db/AL123456_rev.gff\"\n",
    "# Annotation file\n",
    "annotation_file = \"../db/AL123456_rev.gff\"\n",
    "# Reference fasta file\n",
    "reference_file = '../db/h37rv.fasta'\n",
    "\n",
    "######################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import math\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from multiprocessing import Pool\n",
    "\n",
    "from copy import deepcopy\n",
    "\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Data import CodonTable\n",
    "\n",
    "from tqdm import tqdm\n",
    "\n",
    "from sklearn.metrics import f1_score, roc_auc_score, accuracy_score, precision_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3051: DtypeWarning: Columns (1,3,4,5,6,7,8,9,10,25,28,31,34,36,40,41,42,43,44,45,46,47,48,49) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/ipykernel_launcher.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/ipykernel_launcher.py:47: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    }
   ],
   "source": [
    "drug_dict = {\n",
    "    'AMI': 'Amikacin',\n",
    "    'CAP': 'Capreomycin',\n",
    "    'EMB': 'Ethambutol',\n",
    "    'ETH': 'Ethionamide',\n",
    "    'INH': 'Isoniazid',\n",
    "    'KAN': 'Kanamycin',\n",
    "    'MXF': 'Moxifloxacin',\n",
    "    'PZA': 'Pyrazinamide',\n",
    "    'RIF': 'Rifampicin',\n",
    "    'STM': 'Streptomycin'\n",
    "}\n",
    "\n",
    "drugs = drug_dict.keys()\n",
    "\n",
    "nucl_dict = {\n",
    "    'C': 'G',\n",
    "    'G': 'C',\n",
    "    'A': 'T',\n",
    "    'T': 'A'\n",
    "}\n",
    "# Annotation preparation\n",
    "annotation = pd.read_csv(annotation_file, sep='\\t', index_col=None, header=None)\n",
    "annotation = annotation[annotation[2] == 'Gene']\n",
    "annotation[8] = [x.split(';')[0].split()[1] for x in annotation[8].to_numpy()]\n",
    "annotation = annotation.astype({3: int, 4: int})\n",
    "\n",
    "annotation[3] = annotation[3] - 1\n",
    "annotation[4] = annotation[4] - 1\n",
    "\n",
    "# Who catalogue preparation\n",
    "who_catalogue = pd.read_csv(who_catalogue_file, \n",
    "                            sep=';', index_col=None, header=None)\n",
    "who_catalogue = who_catalogue[who_catalogue[51].isin(['1) Assoc w R'])]\n",
    "nucl_who_catalogue = who_catalogue[~(who_catalogue.isna()[3])]\n",
    "nucl_who_catalogue = nucl_who_catalogue.loc[1:]\n",
    "\n",
    "nucl_who_catalogue = nucl_who_catalogue.astype({7: int, 8: int, 9: int, 10: int})\n",
    "nucl_who_catalogue[7] += nucl_who_catalogue[9]\n",
    "nucl_who_catalogue[8] += nucl_who_catalogue[10]\n",
    "nucl_who_catalogue = nucl_who_catalogue[[0, 2, 3, 7, 8, 9, 11]]\n",
    "nucl_who_catalogue.columns = ['drug', 'desc', 'pos', 'pres_who', 'abs_who', 'pres_r_who', 'ppv_who']\n",
    "nucl_who_catalogue['|'] = '|'\n",
    "\n",
    "aa_snps_who_catalogue = who_catalogue[(who_catalogue.isna()[3])]\n",
    "aa_snps_who_catalogue[7] += aa_snps_who_catalogue[9]\n",
    "aa_snps_who_catalogue[8] += aa_snps_who_catalogue[10]\n",
    "aa_snps_who_catalogue = aa_snps_who_catalogue[[0, 2, 3, 7, 8, 9, 11]]\n",
    "aa_snps_who_catalogue.columns = ['drug', 'desc', 'pos', 'pres_who', \n",
    "                                 'abs_who', 'pres_r_who','ppv_who']\n",
    "aa_snps_who_catalogue['|'] = '|'\n",
    "\n",
    "indel_who_catalogue = nucl_who_catalogue[np.array([len(x.split('_')) == 6 for x in nucl_who_catalogue['desc'].values])]\n",
    "nucl_snps_who_catalogue = nucl_who_catalogue[[True if ('ins' not in x) & \\\n",
    "                                              ('del' not in x) else False \\\n",
    "                                              for x in nucl_who_catalogue['desc'].values]]\n",
    "\n",
    "# Form list of nucleotide samples name\n",
    "samples = os.listdir(nucl_data_dir)\n",
    "samples.remove('samples_low_covered.list')\n",
    "samples.remove('samples_filtered.list')\n",
    "samples.remove('combined1.list')\n",
    "samples.remove('combined2.list')\n",
    "samples.remove('combined3.list')\n",
    "samples.remove('combined4.list')\n",
    "samples.remove('combined5.list')\n",
    "samples.remove('filtered_raw_variants_pos.csv')\n",
    "samples.remove('low_covered_raw_variants_pos.csv')\n",
    "\n",
    "# Reference \n",
    "f = open('../db/h37rv.fasta', 'r')\n",
    "reference = \"\".join([x[:-1] if x[-1] == '\\n' else x for x in f.readlines()[1:]])\n",
    "f.close()\n",
    "\n",
    "comp_reference = \"\".join([nucl_dict[x] for x in reference][::-1])\n",
    "\n",
    "LENGHT_GEN = len(reference)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare matrix containing indels information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def converter(mut, pos, reference):\n",
    "    pos = LENGHT_GEN - pos + 1\n",
    "    LEN_SEQ = 100\n",
    "\n",
    "    e = 20\n",
    "    # print(reference[pos-e:pos+3])\n",
    "    # print(\"\".join([nucl_dict[x] for x in reference[pos-e:pos+3]]))\n",
    "\n",
    "    new_reference = deepcopy(reference)\n",
    "    if mut.split('_')[2] == 'del':\n",
    "        fr = mut.split('_')[4].upper()\n",
    "        to = mut.split('_')[5].upper()\n",
    "        \n",
    "        fr_comp_rev = ''.join([nucl_dict[x] for x in list(fr)[::-1]])\n",
    "\n",
    "        if fr_comp_rev[:-1] != reference[pos-len(fr):pos-1]:\n",
    "            print(\"Error: complement substitution is not equal to the reference!\")\n",
    "            \n",
    "        new_reference = reference[:pos-len(fr)] + reference[pos-1:]\n",
    "\n",
    "        return new_reference[pos-len(fr) - LEN_SEQ:pos-len(fr) + LEN_SEQ], \\\n",
    "                reference[pos - LEN_SEQ*2:pos + LEN_SEQ*2], \\\n",
    "                pos - LEN_SEQ*2\n",
    "    else:\n",
    "        fr = mut.split('_')[4].upper()\n",
    "        to = mut.split('_')[5].upper()\n",
    "        \n",
    "        fr_comp_rev = ''.join([nucl_dict[x] for x in list(fr)[::-1]])\n",
    "        to_comp_rev = ''.join([nucl_dict[x] for x in list(to)[::-1]])\n",
    "\n",
    "        # if nucl_dict[fr] != reference[pos-1]:\n",
    "        #     print(\"Error: complement substitution is not equal to the reference!\")\n",
    "        new_reference = reference[:pos-len(fr)] + to_comp_rev + reference[pos:] \n",
    "        \n",
    "        return new_reference[pos - LEN_SEQ:pos + LEN_SEQ], \\\n",
    "                reference[pos - LEN_SEQ*2:pos + LEN_SEQ*2], \\\n",
    "                pos - LEN_SEQ*2\n",
    "\n",
    "def caller(alignment, shift):\n",
    "    seq_1 = alignment[0][0]\n",
    "    seq_2 = alignment[0][1]\n",
    "    \n",
    "    gap_st = -1\n",
    "    gap_end = -1\n",
    "    flag = 0\n",
    "    for i, n in enumerate(seq_1):\n",
    "        if (flag == 0) & (n != '-'):\n",
    "            flag = 1\n",
    "        if (flag == 1) & (n == '-'):\n",
    "            gap_st = i\n",
    "            flag = 2\n",
    "        if (flag == 2) & (n != '-'):\n",
    "            gap_end = i\n",
    "            break\n",
    "    \n",
    "    if gap_end == -1:\n",
    "        min_shift = LENGHT_GEN\n",
    "        min_j = 0\n",
    "        for j in range(0, len(alignment)):\n",
    "            flag = 0\n",
    "            for i, n in enumerate(alignment[j][1]):\n",
    "                if (flag == 0) & (n != '-'):\n",
    "                    flag = 1\n",
    "                if (flag == 1) & (n == '-'):\n",
    "                    if i < min_shift:\n",
    "                        min_shift = i\n",
    "                        gap_st = i\n",
    "                        flag = 2\n",
    "                if (flag == 2) & (n != '-'):\n",
    "                    gap_end = i\n",
    "                    min_j = j\n",
    "                    break\n",
    "        seq_1 = alignment[min_j][0]\n",
    "        seq_2 = alignment[min_j][1] \n",
    "\n",
    "        return [gap_st-1+1+shift, seq_1[gap_st-1:gap_end], seq_1[gap_st-1]]\n",
    "    else:\n",
    "        min_shift = gap_st\n",
    "        min_j = 0\n",
    "        for j in range(1, len(alignment)):\n",
    "            flag = 0\n",
    "            for i, n in enumerate(alignment[j][0]):\n",
    "                if (flag == 0) & (n != '-'):\n",
    "                    flag = 1\n",
    "                if (flag == 1) & (n == '-'):\n",
    "                    if i < min_shift:\n",
    "                        min_shift = i\n",
    "                        gap_st = i\n",
    "                        flag = 2\n",
    "                if (flag == 2) & (n != '-'):\n",
    "                    gap_end = i\n",
    "                    min_j = j\n",
    "                    break\n",
    "        seq_1 = alignment[min_j][0]\n",
    "        seq_2 = alignment[min_j][1] \n",
    "        \n",
    "        return [gap_st-1+1+shift, seq_2[gap_st-1], seq_2[gap_st-1:gap_end]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_mut_converted = []\n",
    "for mut, pos, in tqdm(indel_who_catalogue[['desc','pos']].to_numpy()[:2]):\n",
    "    pos = int(pos)\n",
    "    changed_seq, seq, shift = converter(mut, pos, reference)\n",
    "    alignment = pairwise2.align.globalxs(reference, seq,-1,0)\n",
    "    list_mut_converted.append(caller(alignment, shift))\n",
    "indel_who_catalogue['conv_mut'] = [\"_\".join([str(y) for y in x]) for x in list_mut_converted]\n",
    "indel_who_catalogue.to_csv(\"indel_info.csv\", sep='\\t', index=True, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "indel_who_catalogue = pd.read_csv(\"indel_info.csv\", sep='\\t', index_col=0, header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 12814/12814 [00:19<00:00, 657.24it/s]\n"
     ]
    }
   ],
   "source": [
    "output = dict.fromkeys(indel_who_catalogue['conv_mut'].values)\n",
    "\n",
    "for mut in indel_who_catalogue['conv_mut'].values:\n",
    "    output[mut] = []\n",
    "\n",
    "list_mut_conv_mod = [\"\\t\".join([str(y) for y in x.split('_')]) + '\\n' for x in indel_who_catalogue['conv_mut'].values]\n",
    "    \n",
    "for sample in tqdm(samples):\n",
    "    sample_name = sample.split('.')[0]\n",
    "    f = open(f'{nucl_data_dir}/{sample}', 'r')\n",
    "    for line in f.readlines():\n",
    "        if line in list_mut_conv_mod:\n",
    "            [pos, fr, to] = line.split('\\t')\n",
    "            to = to[:-1]\n",
    "            target = '_'.join([pos, fr, to])\n",
    "            output[target].append(sample_name)\n",
    "    f.close()\n",
    "    \n",
    "filtered_indel_who_catalogue = indel_who_catalogue[indel_who_catalogue.index != 13044]\n",
    "\n",
    "data = pd.DataFrame(0, index=[x.split('.')[0] for x in samples], \n",
    "                      columns=filtered_indel_who_catalogue['conv_mut'].values)\n",
    "for mut in indel_who_catalogue['conv_mut'].values:\n",
    "    data.loc[output[mut], mut] = 1\n",
    "    \n",
    "data.to_csv(\"indel_matrix.csv\", sep='\\t', index=True, header=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare matric containing nucleotide snps information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_strain(annotation, gene):\n",
    "    return annotation[annotation[8] == gene][6].values[0]\n",
    "\n",
    "def convert_who(mut, pos, annotation, reference):\n",
    "    \n",
    "    pos = LENGHT_GEN - int(pos) + 1\n",
    "    \n",
    "    if len(mut.split()) == 2:\n",
    "        mut_1 = mut.split()[0]\n",
    "        mut_2 = mut.split()[1][1:-1]\n",
    "        \n",
    "        pos_1 = int(mut_1.split('_')[1][1:-1])\n",
    "        pos_2 = int(mut_2.split('_')[1][1:-1])\n",
    "        \n",
    "        if abs(pos_1) <= abs(pos_2):\n",
    "            mut = mut_1\n",
    "        else:\n",
    "            mut = mut_2\n",
    "            \n",
    "    gene = mut.split('_')[0]\n",
    "    \n",
    "        \n",
    "    if len(mut.split('_')) == 2:\n",
    "        subst = mut.split('_')[1]\n",
    "        \n",
    "        if return_strain(annotation, gene) == '-':\n",
    "            fr = nucl_dict[subst[0].upper()]\n",
    "            to = nucl_dict[subst[-1].upper()]\n",
    "        else:\n",
    "            fr = subst[0].upper()\n",
    "            to = subst[-1].upper()\n",
    "        \n",
    "        return [pos, fr, to]\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  import sys\n"
     ]
    }
   ],
   "source": [
    "list_of_snps_mut = [convert_who(mut, pos, annotation, reference) for mut, pos \\\n",
    "                    in nucl_snps_who_catalogue[['desc', 'pos']].to_numpy()]\n",
    "list_of_snps_mut = ['_'.join([str(y) for y in x]) for x in list_of_snps_mut]\n",
    "\n",
    "drug_list = [drug_dict[x] for x in nucl_snps_who_catalogue['drug']]\n",
    "\n",
    "nucl_snps_who_catalogue['raw_name'] = list_of_snps_mut\n",
    "\n",
    "nucl_snps_who_catalogue.to_csv(\"nucl_snps_info.csv\", sep='\\t', index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 12814/12814 [00:27<00:00, 470.64it/s]\n"
     ]
    }
   ],
   "source": [
    "output = dict.fromkeys(list_of_snps_mut, '')\n",
    "\n",
    "for sample in tqdm(samples):\n",
    "    sample_name = sample.split('.')[0]\n",
    "    \n",
    "    f = open(f\"{nucl_data_dir}/{sample}\", 'r')\n",
    "    for mut in f.readlines():\n",
    "        mut = \"_\".join(mut[:-1].split('\\t'))\n",
    "        \n",
    "        if mut in list_of_snps_mut:\n",
    "            if output[mut] == '':\n",
    "                output[mut] = [sample_name]\n",
    "            else:\n",
    "                output[mut].append(sample_name)\n",
    "    f.close()\n",
    "    \n",
    "data = pd.DataFrame(0, index=[x.split('.')[0] for x in samples], columns=np.unique(list_of_snps_mut))\n",
    "\n",
    "for mut in output.keys():\n",
    "    indel_data.loc[output[mut], mut] = 1\n",
    "    \n",
    "data.to_csv(\"nucl_snps_matrix.csv\", sep='\\t', index=True, header=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare matric containing aminoacids snps information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_snps_who_catalogue.to_csv(\"aa_snps_info.csv\", sep='\\t', index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['embB', 'ethA', 'gid', 'gyrA', 'gyrB', 'katG', 'pncA', 'rplC',\n",
       "       'rpoB', 'rpsL', 'tlyA'], dtype='<U4')"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = np.unique([x.split('_')[0] for x in aa_snps_who_catalogue['desc']])\n",
    "anno_gene_list = annotation[[8, 3, 4, 6]][annotation[8].isin(gene_list)].to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 12814/12814 [04:19<00:00, 49.30it/s]\n",
      "100%|██████████| 12814/12814 [00:06<00:00, 2003.93it/s]\n"
     ]
    }
   ],
   "source": [
    "result = dict.fromkeys([x.split('.')[0] for x in samples])\n",
    "all_features = set()\n",
    "\n",
    "for sample in tqdm(samples):\n",
    "    sample_name = sample.split('.')[0]\n",
    "    output = []\n",
    "    f = open(f'{nucl_data_dir}/{sample}', 'r')\n",
    "    for line in f.readlines():\n",
    "        if len(line.split('\\t')) == 1:\n",
    "            print(sample)\n",
    "        pos, fr, to = line.split('\\t')\n",
    "\n",
    "        pos = int(pos) - 1\n",
    "        to = to[:-1]\n",
    "\n",
    "        for [gene, st, end, strand] in anno_gene_list:\n",
    "\n",
    "            if (st <= pos) & (pos < end) & (len(fr) == 1) & (len(to) == 1):\n",
    "\n",
    "                gene_seq = reference[st:end+1]\n",
    "                gene_pos = pos - st\n",
    "\n",
    "                codon_number = math.floor(gene_pos / 3)\n",
    "                codon_pos = gene_pos % 3\n",
    "\n",
    "                codon_seq = gene_seq[(codon_number) * 3:(codon_number + 1) * 3]\n",
    "\n",
    "                if codon_seq[codon_pos] != fr:\n",
    "                    print('Error! Substitution is not equal to the reference')\n",
    "\n",
    "                if codon_pos == 0:\n",
    "                    mod_codon_seq = to + codon_seq[codon_pos+1:]\n",
    "                elif codon_pos == 1:\n",
    "                    mod_codon_seq = codon_seq[:codon_pos]+ to + codon_seq[codon_pos+1:]\n",
    "                else:\n",
    "                    mod_codon_seq = codon_seq[:codon_pos] + to\n",
    "\n",
    "                if strand == '-':\n",
    "                    codon_number = int(len(gene_seq)/3) - (codon_number + 1)\n",
    "\n",
    "                    codon_A = Seq(codon_seq).reverse_complement().translate(stop_symbol='!')\n",
    "\n",
    "                    codon_B = Seq(mod_codon_seq).reverse_complement().translate(stop_symbol='!')\n",
    "                else:\n",
    "                    codon_A = Seq(codon_seq).translate(stop_symbol='!')\n",
    "\n",
    "                    codon_B = Seq(mod_codon_seq).translate(stop_symbol='!')\n",
    "\n",
    "                if codon_A == codon_B:\n",
    "                    continue\n",
    "\n",
    "                output.append(f'{gene}_{codon_A}{codon_number+1}{codon_B}')\n",
    "    f.close()\n",
    "    \n",
    "    all_features = all_features | set(output)\n",
    "    result[sample_name] = output\n",
    "    \n",
    "final = pd.DataFrame(0, columns=list(all_features), index=[x.split('.')[0] for x in samples])\n",
    "\n",
    "for sample in tqdm(samples):\n",
    "    sample_name = sample.split('.')[0]\n",
    "    final.loc[sample_name, result[sample_name]] = 1\n",
    "    \n",
    "final.to_csv(\"aa_snps_matrix.csv\", sep='\\t', index=True, header=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_auc(y_test, y_proba):\n",
    "\n",
    "    auc = 0\n",
    "    try:\n",
    "        auc = roc_auc_score(y_test, y_proba)\n",
    "    except ValueError:\n",
    "        pass\n",
    "    return auc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "indel_data = pd.read_csv(\"indel_matrix.csv\", sep='\\t', index_col=0, header=0)\n",
    "nucl_snps_data = pd.read_csv(\"nucl_snps_matrix.csv\", sep='\\t', index_col=0, header=0)\n",
    "aa_data = pd.read_csv('aa_snps_matrix.csv', sep='\\t', index_col=0, header=0)\n",
    "\n",
    "indel_info = pd.read_csv(\"indel_info.csv\", sep='\\t', index_col=None, header=0)\n",
    "nucl_snps_info = pd.read_csv(\"nucl_snps_info.csv\", sep='\\t', index_col=None, header=0)\n",
    "aminoacids_info = pd.read_csv(\"aa_snps_info.csv\", sep='\\t', index_col=None, header=0)\n",
    "\n",
    "data = pd.concat([indel_data, nucl_snps_data, aa_data], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/ipykernel_launcher.py:25: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/pandas/core/series.py:1021: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._where(~key, value, inplace=True)\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/ipykernel_launcher.py:25: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/pandas/core/series.py:1021: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._where(~key, value, inplace=True)\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/ipykernel_launcher.py:25: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/pandas/core/series.py:1021: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._where(~key, value, inplace=True)\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/ipykernel_launcher.py:25: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/pandas/core/series.py:1021: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._where(~key, value, inplace=True)\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/ipykernel_launcher.py:25: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/pandas/core/series.py:1021: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._where(~key, value, inplace=True)\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/ipykernel_launcher.py:25: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/pandas/core/series.py:1021: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._where(~key, value, inplace=True)\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/ipykernel_launcher.py:25: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/pandas/core/series.py:1021: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._where(~key, value, inplace=True)\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/ipykernel_launcher.py:25: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/pandas/core/series.py:1021: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._where(~key, value, inplace=True)\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/ipykernel_launcher.py:25: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/pandas/core/series.py:1021: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._where(~key, value, inplace=True)\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/ipykernel_launcher.py:25: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/export/home/fedonin/reshetnikov/env/lib/python3.6/site-packages/pandas/core/series.py:1021: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._where(~key, value, inplace=True)\n"
     ]
    }
   ],
   "source": [
    "auc = []\n",
    "f1 = []\n",
    "sensitivity = []\n",
    "specificity = []\n",
    "PPV = []\n",
    "NPV = []\n",
    "\n",
    "result = pd.DataFrame(index=[drug_dict[x] for x in drug_dict.keys()])\n",
    "\n",
    "for drug in drugs:\n",
    "    drug_name = drug_dict[drug]\n",
    "    \n",
    "    features = []\n",
    "    \n",
    "    features.extend(indel_info[indel_info['drug'] == drug]['conv_mut'].values)\n",
    "    features.extend(nucl_snps_info[nucl_snps_info['drug'] == drug]['raw_name'].values)\n",
    "    features.extend(aminoacids_info[aminoacids_info['drug'] == drug]['desc'].values)\n",
    "    \n",
    "    pheno = pd.read_csv(f\"{pheno_dir}/{drug_name}.pheno\", sep='\\t', \n",
    "                        index_col=0, header=None, names=['pheno'])\n",
    "    \n",
    "    features = list(set(features) & set(data.columns))\n",
    "    pheno_data = data[features]\n",
    "    \n",
    "    pheno_data['sum'] = np.sum(pheno_data, axis=1)\n",
    "    pheno_data['sum'][pheno_data['sum'] != 0] = 1\n",
    "    \n",
    "    pheno_data = pd.merge(pheno_data['sum'], pheno, left_index=True, right_index=True, how='right')\n",
    "    \n",
    "    auc.append(get_auc(pheno_data['pheno'].values, pheno_data['sum'].values))\n",
    "    f1.append(f1_score(pheno_data['pheno'].values, pheno_data['sum'].values))\n",
    "    \n",
    "    TP = np.sum(np.sum(pheno_data[pheno_data['sum'] == 1], axis=1) == 2)\n",
    "    FP = np.sum(np.sum(pheno_data[pheno_data['sum'] == 1], axis=1) == 1)\n",
    "    TN = np.sum(np.sum(pheno_data[pheno_data['sum'] == 0], axis=1) == 0)\n",
    "    FN = np.sum(np.sum(pheno_data[pheno_data['sum'] == 0], axis=1) == 1)\n",
    "    \n",
    "    sensitivity.append(TP/(TP + FN))\n",
    "    specificity.append(TN/(TN + FP))\n",
    "    PPV.append(TP/(TP + FP))\n",
    "    NPV.append(TN/(TN + FN))\n",
    "    \n",
    "result['auc'] = auc\n",
    "result['sensitivity'] = sensitivity\n",
    "result['specificity'] = specificity\n",
    "result['PPV'] = PPV\n",
    "result['NPV'] = NPV\n",
    "\n",
    "result.to_csv('TableS23', sep='\\t', index=True, header=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
